### PR TITLE
Add myopic optimization

### PR DESCRIPTION
**Pull request by davide-f**
# Closes # (if applicable).

## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [x] Code and workflow changes are sufficiently documented.
- [x] Newly introduced dependencies are added to `envs/environment.yaml` and `envs/environment.docs.yaml`.
- [x] Changes in configuration options are added in all of `config.default.yaml`, `config.tutorial.yaml`, and `test/config.test1.yaml`.
- [x] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [x] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes, including reference to the requested PR.

### CODE REVIEW

**Comment by davide-f** [comment]
This PR builds on #285 , priority shall be given to #285

**Comment by davide-f** [comment]
@energyLS FYI :)

**Comment by energyLS** [comment]
I will take over and review this PR

File: config.default.yaml
Comment (energyLS) [inline config.default.yaml:166]: Can you elaborate what the effect of "grouping_years_power" is and what values are necessary/allowed?

File: data/existing_infrastructure/existing_heating_raw.csv
Comment (energyLS) [inline data/existing_infrastructure/existing_heating_raw.csv:31]: No global data, is this an issue? Are there default fallback values?

File: scripts/add_brownfield.py
Comment (energyLS) [inline scripts/add_brownfield.py:179]: Why is this not required?

File: scripts/add_export.py
Comment (energyLS) [inline scripts/add_export.py:80]: I see the purpose, does "world" make sense?

File: scripts/prepare_sector_network.py
Comment (energyLS) [inline scripts/prepare_sector_network.py:2440]: When should this be moved to pypsa-earth?

File: scripts/prepare_sector_network.py
Comment (energyLS) [inline scripts/prepare_sector_network.py:2464]: Planning for perfect optimisation, I like the optimism ;)

**Review by energyLS — COMMENTED** [review]
Great job. I did not review "add_existing_baseyear.py" and "add_brownfield.py" yet. Currently trying to run it locally.

File: Snakefile
Comment (davide-f) [reply Snakefile:18]: Yes, basically it loads each config file and for any new config, the changed parameters are overwritten.
This may allow the config.yaml to be even  lighter and keep only the needed changes.
The goal of this is to have all pypsa-earth parameters within the simulation and will also help the merge as there is no need for duplication.

In some scripts, some pypsa-earth parameters were needed but that required the manual adjustment of the config.pypsa-earth.yaml and config.yaml independently.
With this is no more needed: the parameter shall stay in config.pypsa-earth
For context, in pypsa-eur they now moved to a similar setup.

In the past, we considered this solution in pypsa-earth but the "merge" of configs was not smooth as it is now

File: Snakefile
Comment (davide-f) [reply Snakefile:482]: Shallow execution requires the rule to be run in a separate temporary directory.
The value regulates how the execution is performed and how the input/output data are treated between the temporary and the main folder.
see: https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#shadow-rules

File: Snakefile
Comment (davide-f) [reply Snakefile:629]: Yes

File: Snakefile
Comment (davide-f) [reply Snakefile:701]: This rule is executed only for the first network and therefore sets the inputs for the first year planning horizon

File: Snakefile
Comment (davide-f) [reply Snakefile:801]: It feels like the rules must be defined before adding this line, we could test, but it is likely that it does not work

File: data/existing_infrastructure/existing_heating_raw.csv
Comment (davide-f) [reply data/existing_infrastructure/existing_heating_raw.csv:31]: Good point, to add

File: scripts/add_export.py
Comment (davide-f) [reply scripts/add_export.py:80]: Other recommedation?
Earth?

I would avoid something customizable that is likely an unnecessary complication, if you agree too

File: scripts/prepare_sector_network.py
Comment (davide-f) [reply scripts/prepare_sector_network.py:2440]: After the merge?
I believe the sector script requires some heavy changes and there is the need to update the environment too.
If you agree I open an issue on that

File: scripts/prepare_sector_network.py
Comment (davide-f) [reply scripts/prepare_sector_network.py:2464]: ahahah

File: scripts/add_brownfield.py
Comment (davide-f) [reply scripts/add_brownfield.py:179]: This optionally updates the input renewable time series for the components across years.
When I tested, it lead to problems and I considered of secondary priority but worth keeping to remind of future updates.
What do you think?

File: Snakefile
Comment (davide-f) [reply Snakefile:723]: This is a trick to enforce the rule to be executed only for the first planning horizon.
wildcard_constraints enforces a constraint on the wildcard that ensures:
planning_horizons=config["scenario"]["planning_horizons"][0]

Thus only the first horizon is executed

**Comment by energyLS** [comment]
Testing locally and adding 2050 as well, I run into this error. Currently investigating.
grafik

**Comment by davide-f** [comment]
Thanks! Leave it to me :)
I'd recommend to test: 2030 and 2050 as horizons (I tested them)

**Comment by energyLS** [comment]
> Thanks! Leave it to me :) I'd recommend to test: 2030 and 2050 as horizons (I tested them)

Thanks, I will leave that to you

**Comment by energyLS** [comment]
however good news: Apart from that error in 2050, everything seems to work!

File: Snakefile
Comment (energyLS) [reply Snakefile:723]: Nice, great trick!

File: data/existing_infrastructure/existing_heating_raw.csv
Comment (energyLS) [reply data/existing_infrastructure/existing_heating_raw.csv:31]: Are you adding something or how should we proceed here?

File: scripts/add_brownfield.py
Comment (energyLS) [reply scripts/add_brownfield.py:179]: Ah I see, interesting point. How is it handled currently, if I do a overnight scenario for 2030 and an overnight for 2050, the profiles are the same, right? If that's the case, we don't have to implement this feature here now but leave it for future developments.

File: scripts/add_export.py
Comment (energyLS) [reply scripts/add_export.py:80]: Earth sounds great, let's do that

File: scripts/prepare_sector_network.py
Comment (energyLS) [reply scripts/prepare_sector_network.py:2440]: I agree, please feel free to open an issue on that and then resolve this conversation

File: data/existing_infrastructure/existing_heating_raw.csv
Comment (davide-f) [reply data/existing_infrastructure/existing_heating_raw.csv:31]: Added a DEFAULT line similarly to the demand folder.
When DEFAULTs are loaded, the logging triggers that.
To do so a new rule is also added

File: scripts/add_export.py
Comment (davide-f) [reply scripts/add_export.py]: Done

File: scripts/prepare_sector_network.py
Comment (davide-f) [reply scripts/prepare_sector_network.py:2440]: Done

File: scripts/add_brownfield.py
Comment (davide-f) [reply scripts/add_brownfield.py:179]: Exactly, that's the point. That's why it is also kept commented.
I resolve this comment, feel free to reopen it if appropriate :)

File: config.default.yaml
Comment (davide-f) [reply config.default.yaml:166]: Assets are installed with "continuous" years, e.g. 1960, 1961, 1962 etc.
The info are parsed and grouped at custom "bins" years like histogram and this option grouping years specify the "bins".
This is then used to clarify the lifetime of the components in a more "discrete" approach according to the modeler will to define the planning_horizons.
Not sure if that's clear

File: Snakefile
Comment (davide-f) [reply Snakefile:770]: Apologies this comment skipped.
I didn't put too much attention here as the inputs profile_* are used in the function "adjust_renewable_profiles" that is commented.
I agree that extending this to hydro is also very interesting but requires some code revision as the modelling of hydro is different from solar and wind.

Not sure if it is worth doing it here however, also considering that the function is commented

File: Snakefile
Comment (energyLS) [reply Snakefile:726]: @davide-f do you have an idea on that?

File: Snakefile
Comment (energyLS) [reply Snakefile:770]: alright, that is fine not to integrate it in this PR

File: Snakefile
Comment (energyLS) [reply Snakefile:824]: That's fine then, let's keep it the way it is

File: config.default.yaml
Comment (energyLS) [reply config.default.yaml:166]: It is more or less clear yes, how should we handle this information? Small comment in the config, since the docs are not activate for earth-sec

File: data/existing_infrastructure/existing_heating_raw.csv
Comment (energyLS) [reply data/existing_infrastructure/existing_heating_raw.csv:31]: Great, thanks.

**Comment by energyLS** [comment]
@davide-f do you require any additional support here?

File: Snakefile
Comment (davide-f) [reply Snakefile:809]: Oh sorry, yeah that is correct. It applies to components that have no build_year; those that do have it are not affected by this option. I'm investigating the 2040 case and verify there

**Comment by davide-f** [comment]
This PR is ready to go and contains the fix for DF 2040 too.
It can be checked with https://github.com/pypsa-meets-earth/documentation/pull/89

File: Snakefile
Comment (davide-f) [reply Snakefile:809]: The capacity expansion works well also for the 2040 case. The notebook can be used to crosscheck

**Review by energyLS — APPROVED** [review]
Approved. Great job @davide-f!

File: test/config.test1.yaml
Comment (energyLS) [inline test/config.test1.yaml:122]: @finozzifa see here for newly introduced config option

**Pull request merged**
